### ISSUE TITLE
[java] UseArraysAsList false negative

### ISSUE THREAD

**Issue by maikelsteneker** [pmd/pmd#4014]
**Affects PMD Version:** 6.46.0

**Rule:** [UseArraysAsList](https://pmd.github.io/latest/pmd_rules_java_performance.html#usearraysaslist)

**Description:** When upgrading from PMD 6.33.0, we found a false negative.

**Code Sample demonstrating the issue:**

```java
/** Like Array.asList() but with varargs and returns a mutable list
  *
  * @param elements    Array of elements or number of elements to add
  *                    in a new list. If the only given parameter is
  *                    null, no elements will be added and an empty
  *                    list returned. If more parameters, null values
  *                    will be added in the list. 
  * @return            List holding elements
  */
public static <T> List<T> asList(T ... elements) {
    final int size = isEmpty(elements) ? 0 : elements.length;
    final List<T> list = new ArrayList<T>( /*initialCapacity=*/ size + 10 );
    if( size > 0 ) {
        for(int i=0; i<size; i++) list.add( elements[i] );
    }
    return list;
}
```

**Expected outcome:** The `list.add` invocation should result in a violation, but it doesn't. We're running through the CLI without type information; could that be related?

**Running PMD through:** *CLI*

**Comment by jsotuyod** [comment]
This is actually down to the usage of a variable (size) and not directly elements.length. The XPath is explicit about it, as we don't track the value of variables.

Relaxing this condition would fix this, but introduce new FPs in cases of partial copies, where asList would not be a proper replacement.

I don't think PMD 7 will cope any better with this unfortunately.

**Comment by jsotuyod** [comment]
This is no longer reproducible under PMD 7.0.0
